File: ios/Sources/KlinikAPI/RecordSubject.swift
import Foundation

/**
 * Whose records a request is about.
 *
 * The same distinction `MeasurementSubject` already draws, and for the same
 * reason it is a type rather than an optional id: the two paths are different
 * on the server, because the permission is. A patient has `self.read`, not
 * `documents.read` — giving them the latter would bring every patient's file
 * within reach, which is exactly what the permission matrix exists to prevent.
 *
 * So `me/documents` is not a convenience alias for `patients/{id}/documents`.
 * Sending a patient down the staff path answers 403, and this type is what
 * stops a screen being wired to the wrong one by accident.
 */
public enum RecordSubject: Sendable, Equatable {
    case patient(id: String)
    case me

    /// The path prefix for this subject, e.g. `me` or `patients/<id>`.
    func base(_ suffix: String) -> String {
        switch self {
        case .patient(let id): return "patients/\(id)/\(suffix)"
        case .me: return "me/\(suffix)"
        }
    }
}

File: android/core/network/src/main/kotlin/xyz/klinik/network/RecordSubject.kt
package xyz.klinik.network

/**
 * Whose records a request is about.
 *
 * A type rather than a nullable id, because the two paths differ on the server
 * and the reason is the permission: a patient holds `self.read`, not
 * `documents.read`. Handing a patient the staff path answers 403 — and it did,
 * for four of their own screens, until the `me/` routes existed.
 *
 * The type is what stops a screen being wired to the wrong one: getting it
 * wrong is a compile error here rather than a permission error a patient sees
 * when they tap a tile.
 */
sealed interface RecordSubject {
    data class Patient(val id: String) : RecordSubject
    data object Me : RecordSubject

    /** The path for this subject, e.g. `me/documents` or `patients/<id>/documents`. */
    fun base(suffix: String): String = when (this) {
        is Patient -> "patients/$id/$suffix"
        Me -> "me/$suffix"
    }
}
